File: golang/matrix/game_of_life.go
package main

func gameOfLife(board [][]int) {
	rows, cols := len(board), len(board[0])

	countLiveNeighbors := func(r, c int) int {
		directions := [][2]int{{-1, -1}, {-1, 0}, {-1, 1}, {0, -1}, {0, 1}, {1, -1}, {1, 0}, {1, 1}}

		count := 0

		for _, dir := range directions {
			nr, nc := r+dir[0], c+dir[1]

			if nr >= 0 && nr < rows && nc >= 0 && nc < cols && abs(board[nr][nc]) == 1 {
				count++
			}
		}

		return count
	}

	for r := 0; r < rows; r++ {
		for c := 0; c < cols; c++ {
			liveNeighbors := countLiveNeighbors(r, c)

			if board[r][c] == 1 && (liveNeighbors < 2 || liveNeighbors > 3) {
				board[r][c] = -1
			}

			if board[r][c] == 0 && liveNeighbors == 3 {
				board[r][c] = 2
			}
		}
	}

	for r := 0; r < rows; r++ {
		for c := 0; c < cols; c++ {
			if board[r][c] > 0 {
				board[r][c] = 1
			} else {
				board[r][c] = 0
			}
		}
	}
}

func abs(x int) int {
	if x < 0 {
		return -x
	}

	return x
}

File: python/matrix/game_of_life.py
from typing import List


class Solution:
    def gameOfLife(self, board: List[List[int]]) -> None:
        rows, cols = len(board), len(board[0])

        def count_live_neighbors(r: int, c: int) -> int:
            directions = [
                (-1, -1),
                (-1, 0),
                (-1, 1),
                (0, -1),
                (0, 1),
                (1, -1),
                (1, 0),
                (1, 1),
            ]
            count = 0
            for dr, dc in directions:
                nr, nc = r + dr, c + dc
                if 0 <= nr < rows and 0 <= nc < cols and abs(board[nr][nc]) == 1:
                    count += 1
            return count

        for r in range(rows):
            for c in range(cols):
                live_neighbors = count_live_neighbors(r, c)

                if board[r][c] == 1 and (live_neighbors < 2 or live_neighbors > 3):
                    board[r][c] = -1

                if board[r][c] == 0 and live_neighbors == 3:
                    board[r][c] = 2

        for r in range(rows):
            for c in range(cols):
                if board[r][c] > 0:
                    board[r][c] = 1
                else:
                    board[r][c] = 0
